pkg/urbit/vere/io/ames.c

```diff
@@ -232,6 +302,99 @@ u3_ames_encode_lane(u3_lane lan) {
   return u3ke_jam(u3nt(c3__ipv4, u3i_words(1, &lan.pip_w), lan.por_s));
 }
 
+/* _ames_lane_from_sockaddr(): sockaddr_in to lane noun
+*/
+static u3_noun
+_ames_lane_from_sockaddr(struct sockaddr_in* add_u)
+{
+  u3_lane lan_u;
+  lan_u.por_s = ntohs(add_u->sin_port);
+  lan_u.pip_w = ntohl(add_u->sin_addr.s_addr);
+  return u3_ames_encode_lane(lan_u);
+}
+
+/* _ames_serialize_packet(): u3_panc to atom, updating the origin lane if dop_o
+**                           (retains pac_u)
+*/
+static u3_noun
+_ames_serialize_packet(u3_panc* pac_u, c3_o dop_o)
+{
+  c3_y sen_y = 2 << pac_u->hed_u.sac_y;
+  c3_y rec_y = 2 << pac_u->hed_u.rac_y;
+  c3_o nal_o = c3n;
+
+  //  update the body's lane, if desired
+  //
+  if (c3y == dop_o) {
+    //  unpack (jam [(unit lane) body])
+    //
+    u3_noun lon, bod;
+    {
+      u3_noun old = u3qe_cue(pac_u->bod_u.con);
```

Comment:
This should be virtualized (with `u3m_soft()`), or invalid input will bring down the process. Inability to cue signifies bad input, so this should happen at a point where failure can guide control flow, perhaps inside the conditionals of `_ames_recv_cb()`.